In [1]:
#importing the libraries for numpy, pandas, seaborn, warnings
import numpy as np
import pandas as pd
import seaborn as sns
import warnings

#importing the libraries for svm, SVC, GridSearchCV, pyplot, metrics
from sklearn import svm
from sklearn.svm import SVC
from sklearn import metrics
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split, GridSearchCV

warnings.filterwarnings('ignore')
%matplotlib inline

In [2]:
#print multiple statements in same line
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity ="all"

In [3]:
#load iris dataset 
iris = sns.load_dataset('iris')
iris.head()
iris.shape
iris.info()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

(150, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal_length    150 non-null float64
sepal_width     150 non-null float64
petal_length    150 non-null float64
petal_width     150 non-null float64
species         150 non-null object
dtypes: float64(4), object(1)
memory usage: 5.9+ KB


In [4]:
#Do Dummy Coding for iris data set 
for col in iris.columns:
    if iris[col].dtype =='object':
        iris[col]=pd.Categorical(iris[col]).codes
    
iris.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

In [5]:
#Create Features and Labels
x = iris.drop('species', axis=1)
y=iris['species']

In [6]:
#Create train and test 70% and 30% split
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.30, random_state=1)
X_train.shape
X_test.shape
y_train.shape
y_test.shape
type(X_train)

(105, 4)

(45, 4)

(105,)

(45,)

pandas.core.frame.DataFrame

## Hyper-parameters Optimization using Gridsearch and Crossvalidation

Support vector machines (SVMs) are a set of supervised learning methods used for classification, 
regression and outliers detection.

In [7]:
svc = SVC()

#Get parameters for this estimator.
svc.get_params()

{'C': 1.0,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 'auto_deprecated',
 'kernel': 'rbf',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

param_grid: specifies that two grids should be explored: one with a linear kernel and C values in [1, 10, 100, 1000], 
and the second one with an RBF kernel, and the cross-product of C values ranging in [1, 10, 100, 1000] 
and gamma values in [0.001, 0.0001].

In [8]:
param_grid = [{'C':[0.01,1,10,100,1000],
              'kernel':['linear','poly','rbf','sigmoid'],
              'gamma':[0.01,1,10,100,200]}]

The grid search provided by GridSearchCV exhaustively generates candidates from a grid of parameter
values specified with the param_grid parameter.

In [9]:
model = GridSearchCV(svc, param_grid, cv=10, scoring='accuracy')
model.fit(x, y)

# Parameter setting that gave the best results on the hold out data.best_params_
model.best_params_

# Mean cross-validated score of the best_estimator
model.best_score_

GridSearchCV(cv=10, error_score='raise-deprecating',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid=[{'C': [0.01, 1, 10, 100, 1000], 'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'gamma': [0.01, 1, 10, 100, 200]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

{'C': 10, 'gamma': 0.01, 'kernel': 'linear'}

0.98

The score of mean is more 98% seems to be good.